pkg/cache/cluster.go

```diff
@@ -553,6 +553,10 @@ func (c *clusterCache) listResources(ctx context.Context, resClient dynamic.Reso
 		}
 
 		listRetry.Steps = int(c.listRetryLimit)
+		// We set the resource version to 0 below to proactively prevent the
+		// list API call from reaching etcd and make the server fetch the data
+		// from the watch cache instead.
+		opts.ResourceVersion = "0"
```

Comment:
I don't think it's *ever* a good idea to list with resourceVersion=0. Even though this allows the API server to use it's object cache, it causes a worse problem. It does not allow limit/pagination/chunking to be used **at all** and so all objects of that group/kind will be returned in the response, resulting in OOM errors.

We've been wrestling with this problem in workflows over the years, and the solution we landed on is to never allow resourceVersion=0 requests, even if it means hitting etcd.

https://github.com/argoproj/argo-workflows/pull/13466#pullrequestreview-2241541998

Reply:
> And do we run the risk of causing massing kube-apiserver memory load by using the watch cache instead of etcd?

>> I'd rather call it trading off some amount of memory for large amount of CPU with the latter being a scarcer resource at scale of large Kubernetes clusters in general :). 

The tradeoff is not really about increase of kube-apiserver CPU/memory. This will cause the **_application-controller_** to have to load all objects of every group/kind in memory when dealing with the non-paginated responses.

Reply:
>The tradeoff is not really about increase of kube-apiserver CPU/memory. This will cause the application-controller to have to load all objects of every group/kind in memory when dealing with the non-paginated responses.

This tradeoff I mentioned was from the server's POV, not client's but I agree there's a risk of client OOMing if the amount of resources it tracks is large and the client doesn't get enough memory assigned. [Transforming informers](https://github.com/kubernetes/kubernetes/pull/104300) could be a possible alleviation of the problem you describe as it reduce the memory footprint due to the objects stored in the client's cache but it would require considerably more work to validate them and migrate to them here if feasible.